src/libexpr/primops.cc

```diff
@@ -2472,6 +2472,76 @@ static RegisterPrimOp primop_mapAttrs({
     .fun = prim_mapAttrs,
 });
 
+static void prim_zipAttrsWith(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    // we will first count how many values are present for each given key.
+    // we then allocate a single attrset and pre-populate it with lists of
+    // appropriate sizes, stash the pointers to the list elements of each,
+    // and populate the lists. after that we replace the list in the every
+    // attribute with the merge function application. this way we need not
+    // use (slightly slower) temporary storage the GC does not know about.
+
+    std::map<Symbol, size_t> attrOccurences;
+    std::map<Symbol, Value * *> lists;
+
+    state.forceFunction(*args[0], pos);
+    state.forceList(*args[1], pos);
+    const auto nrLists = args[1]->listSize();
+
+    for (unsigned int n = 0; n < nrLists; ++n) {
+        Value * vElem = args[1]->listElems()[n];
+        try {
+            state.forceAttrs(*vElem);
+            for (auto & attr : *vElem->attrs)
+                attrOccurences[attr.name]++;
+        } catch (TypeError &e) {
+            e.addTrace(pos, hintfmt("while invoking '%s'", "zipAttrsWith"));
+            throw;
+        }
+    }
+
+    state.mkAttrs(v, attrOccurences.size());
+    for (auto [sym, size] : attrOccurences) {
+        Value * list = state.allocAttr(v, sym);
+        state.mkList(*list, size);
+        lists[sym] = list->listElems();
+    }
+
+    for (unsigned int n = 0; n < nrLists; ++n) {
+        Value * vElem = args[1]->listElems()[n];
+        for (auto & attr : *vElem->attrs)
+            *lists[attr.name]++ = attr.value;
+    }
+
+    for (auto & attr : *v.attrs) {
+        Value * name = state.allocValue();
+        mkString(*name, attr.name);
+        Value * call1 = state.allocValue();
+        mkApp(*call1, *args[0], *name);
+        Value * call2 = state.allocValue();
+        mkApp(*call2, *call1, *attr.value);
+        attr.value = call2;
+    }
+
+    v.attrs->sort();
+}
+
+static RegisterPrimOp primop_zipAttrsWith({
+    .name = "__zipAttrsWith",
+    .args = {"f", "list"},
+    .doc = R"(
+      Merge sets of attributes and use the function f to merge attributes
+      values.
+
+      ```nix
+      builtins.zipAttrsWith (name: vs: vs) [{a = "x";} {a = "y"; b = "z";}]
```

Comment:
```suggestion
      builtins.zipAttrsWith (name: vs: { inherit name vs; }) [{a = "x";} {a = "y"; b = "z";}]
```

This would make for a more complete example, if a little more verbose. (also update value below)